### PR TITLE
Eliminate unnecessary remote exchange before table writer

### PR DESCRIPTION
**Pull request by shixuan-fan**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 1]
```diff
@@ -443,9 +443,6 @@ public PlanNode visitTableScan(TableScanNode node, RewriteContext<FragmentProper
         @Override
         public PlanNode visitTableWriter(TableWriterNode node, RewriteContext<FragmentProperties> context)
         {
-            if (node.getPartitioningScheme().isPresent()) {
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: As i understand that was always extra? Right?

**Comment by arhimondr** [comment]
Failures seem to be related

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java (hunk 1 above)
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: From the message of the commit that introduced this method, it seems to be a sanity check assuming that we won't change the fragment distribution before table writer (https://github.com/prestodb/presto/commit/2323e95ef15ce4d97e441f5b6003e0fa6c931cb4), which is no longer true. This makes me wonder if we should just remove this `visitTableWriter` method entirely.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: > This makes me wonder if we should just remove this visitTableWriter method entirely.

Yeah, let's remote it altogether.

**Comment by shixuan-fan** [comment]
The failure is due to that we should add remote exchange if either of the following hold true:
1. The bucket schema is not compatible (i.e. the target table bucket schema is not super set)
2. Partial merge push down cannot be applied (i.e. bucket count mismatch)

The logic that failed test is broken because we require both of them to be true to add a remote exchange, which is fixed in the updated commit.

**Comment by shixuan-fan** [comment]
I made some hack that makes the test pass, but that definitely should not be the code to be checked in.

The current SPI does not work neatly with super set check, and we might need to rethink SPI design for partitioning-compatibility related issues. The way it works today (https://github.com/prestodb/presto/blob/1e3f420e67b5de8fae9b9f407e7f80d94ed508d5/presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java#L574-L576):
```
1. source partitioning is compatible (to-be-deprecated as annotated)
   with the target table partitioning?
  * yes -> no remote exchange
  * no -> to 2
2. source partitioning is refined partition over target table partitioning 
   and we can push down partial merge?
  * yes -> to 3
  * no -> add remote exchange
3. can push down partial merge?
  * yes -> no remote exchange
  * no -> add remote exchange
```
Yet, both compatible and refined partition check currently return no for both the predicates above when the partitioning schema is different. 

Ideally, the logic should be something like:
```
1. target table partitioning schema is a super set of source partitioning schema?
  * yes -> to 2
  * no -> add remote exchange
2. source partitioning count is compatible with target table partitioning count?
  * yes -> no remote exchange
  * no -> to 3
3. source partitioning is refined partition over target table partitioning 
   and we can push down partial merge?
  * yes -> to 4
  * no -> add remote exchange
4. can push down partial merge?
  * yes -> no remote exchange
  * no -> add remote exchange
```
Ideally we could just make one SPI call, and then based on the return result we could finalize the decision of adding remote exchange or not, or we still need to do step 4.

cc @arhimondr @highker @wenleix thoughts?

**Comment by shixuan-fan** [comment]
@highker: Ummmm. I was trying to ask for thoughts/suggestions for https://github.com/prestodb/presto/pull/13091#issuecomment-513345738. 😂 Currently to make it work I have to add that `hack` commit (https://github.com/prestodb/presto/pull/13091/commits/337b717a7add38baaf410d3ee43d49233e5b564e), which is not ideal.

**Comment by highker** [comment]
@shixuan-fan, I think we have chatted offline regarding that. Have you chatted with @wenleix or @arhimondr on that (unifying the bucketing interface)?

**Comment by shixuan-fan** [comment]
Offline discussed with @wenleix and we figured this remote exchange is **necessary**. Table writer is different from other operations (for example, window function):
- For window function, we only need to care if the required data is already on the node, and can do local exchange to make the data partitioned on the right stream. 
- For TableWriterOperator, we have to make sure we write data into the right bucket. That being said, if source is node partitioned on (A), and the target table requires partitioning on (A, B), then after local exchange, we will have multiple streams trying to write to different buckets, and effectively for each bucket file, there will be multiple threads trying to write to it. The remote exchange is required so that we could distribute the file based on `hash(A, B)` instead of `hash(A)`, and thus each lifespan only needs to write to one file.

**Comment by arhimondr** [comment]
That is tricky. 

In other words:

Let's say that `hash('foo', `baz`) = 0xAABBCCDD` and `hash('oof', `zab`) = 0xAABBCCDD` have the same hash value, thus these two tuples belong to the same bucket. 

When hashing only on the first column, these two tuples may get to two different nodes. It is totally fine for aggregations, as for aggregations the requirement is that same values have to be on the same nodes. When writing a bucketed table the requirement is that all the values that hash to the same hash end up being on the same node. Thus the remote exchange is required.

**Pull request closed without merging**